src/Compilers/CSharp/Portable/Parser/SlidingTextWindow.cs

```diff
@@ -374,6 +374,26 @@ public char PeekChar(int delta)
             return ch;
         }
 
+        public char PreviousChar()
+        {
+            Debug.Assert(this.Position > 0);
+            var desiredPosition = this.Position - 1;
+            if (desiredPosition >= _basis && _offset > 1)
+            {
+                // _basis describes where in the source text the current chunk of characters starts at.
+                // So if the desired position is ahead of that, then we can just read the value out of
+                // the character window directly.
+                return this.CharacterWindow[_offset - 1];
+            }
+
+            // The prior character isn't in the window (trying to read the current character caused us to
+            // read in the next chunk of text into the window, throwing out the preceding characters).
+            // Just go back to the source text to find this character.  While more expensive, this should
+            // be rare given that most of the time we won't be calling this right after loading a new text
+            // chunk.
+            return this.Text[desiredPosition];
```

Comment:
Note: we could *also* simplify the entire method to just this call.  However, it means that any time we see `.num` we will call back into the source-text.  That was a bit concerning to me in the case where maybe there were tons of `.1`-like numbers in a file, and we were having to call all the way back through the source-text for each.

But if we have an easy way to validate performance against a large code-base, we could just try that and go with the much simpler route if we don't see a perf problem.  it would certainly make this method trivial to verify :)